xbmc/filesystem/FileCache.cpp

```diff
@@ -312,25 +370,53 @@ void CFileCache::Process()
       }
     }
 
-    const int64_t maxWrite = m_pCache->GetMaxWriteSize(m_chunkSize);
-    int64_t maxSourceRead = m_chunkSize;
+    const int64_t maxWrite = m_pCache->GetMaxWriteSize(buffer.size());
+    // Make maxSourceRead a multiple of the chunk size, bounded by maxWrite
+    // and m_slowStartReadRate.
+    // If we are falling behind we will use larger and larger reads which
+    // over network file systems will hopefully increase throughput.
+    // Special case: m_chunkSize == 1 (non-buffered file) fits
+    // right into this calculation.
+    int64_t maxSourceRead = std::min(maxWrite, m_slowStartReadSize);
+
+    // Allow a non-m_chunkSize read if only the last chunk remains,
+    // otherwise we won't read to the end.
+    if (m_fileSize == 0 || m_fileSize - m_writePos >= m_chunkSize)
+      // *not* near the end of the file, so round down to nearest chunk size
+      maxSourceRead -= maxWrite % m_chunkSize;
+
     // Cap source read size by space available between current write position and EOF
     if (m_fileSize != 0)
       maxSourceRead = std::min(maxSourceRead, m_fileSize - m_writePos);
 
     /* Only read from source if there's enough write space in the cache
      * else we may keep disposing data and seeking back on (slow) source
      */
-    if (maxWrite < maxSourceRead)
+    if (maxSourceRead <= 0)
     {
       // Wait until sufficient cache write space is available
       m_pCache->m_space.Wait(5ms);
       continue;
     }
 
-    ssize_t iRead = 0;
-    if (maxSourceRead > 0)
-      iRead = m_source.Read(buffer.get(), maxSourceRead);
+    auto avail = m_pCache->WaitForData(0, 0);
+    // If we have less than 1/4 second worth of data left, or if we're about
+    // to read more than what is available, treat that as an early warning
+    // for underrun and log some diagnostic data.
+    // Also log during slow start.
+    if (avail * 4 < m_writeRate || avail < maxSourceRead || m_slowStartReadSize < buffer.size())
+      CLog::Log(LOGDEBUG,
+                "CFileCache::{} Reading {} bytes from source. maxWrite: {}, avail: {}, writeRate: {}, slowStartReadSize: {}",
+                __FUNCTION__, maxSourceRead, maxWrite, avail, m_writeRate, m_slowStartReadSize);
+
+    assert(buffer.size() >= maxSourceRead);
+
+    // Update m_slowStartReadSize for the next round. Once it has reached
+    // buffer.size() it has served it's purpose.
+    if (static_cast<size_t>(m_slowStartReadSize) < buffer.size())
+	m_slowStartReadSize *= 2;
```

Comment:
this multiplier would be a candidate for an advanced setting to allow user tuning.
possibly also read limit.